### PR TITLE
[ADP-322] Add metadata encryption to HTTP API.

### PR DESCRIPTION
**Pull request by paweljakubas**
- [x] Adding  noninvasive api changes to `constructTransaction`
- [x] Applying changes to Api.Server 
- [x] Extend `ErrConstructTx` 
- [x] Implementing `toMetadataEncryption` according to https://github.com/cardano-foundation/CIPs/tree/master/CIP-0083
- [x] Prepare goldens in line to openSSL and CIP83 examples
- [x] Show error emitting works
- [x] Add salting to key gen and encryption
- [x] Add integration test showing error 
- [x] Add integration test showing encryption

### Issue Number
adp-322

### CODE REVIEW

File: lib/crypto-primitives/src/Cryptography/Cipher/AES256CBC.hs [hunk 1]
```diff
@@ -163,3 +164,13 @@ decrypt mode key iv msg = do
     unpad p = case mode of
         WithoutPadding -> Right p
         WithPadding -> maybeToEither EmptyPayload (PKCS7.unpad p)
+
+getSaltFromEncrypted :: ByteString -> Maybe ByteString
+getSaltFromEncrypted msg = do
+    when (BS.length msg < 32) Nothing
+    let (prefix,rest) = BS.splitAt 8 msg
+    let saltDetected = prefix == saltPrefix
```
Comment (jonathanknowles) [inline lib/crypto-primitives/src/Cryptography/Cipher/AES256CBC.hs]: The combination of `splitAt` with `== saltPrefix` can be replaced with `stripPrefix`.

See: https://github.com/cardano-foundation/cardano-wallet/pull/4581

File: lib/api/src/Cardano/Wallet/Api/Types.hs [hunk 2]
```diff
@@ -1237,8 +1238,23 @@ data ApiMultiDelegationAction
...
+instance FromJSON EncryptMetadataMethod where
+    parseJSON = withText "base" $ \txt ->
+        if txt == "basic" then
+            pure AES256CBC
+        else
+            fail "'basic' is expected."
+
+data ApiEncryptMetadata = ApiEncryptMetadata
+    { passphrase :: ApiT (Passphrase "lenient")
+    , enc :: Maybe EncryptMetadataMethod
```
Comment (jonathanknowles) [inline lib/api/src/Cardano/Wallet/Api/Types.hs]: Hi @paweljakubas 

For what reason do we allow this field to be `Nothing`?

If it is set to `Nothing`, then what do we expect the behaviour to be?

File: lib/api/src/Cardano/Wallet/Api/Types.hs [hunk 3]
```diff
@@ -1237,8 +1238,23 @@ data ApiMultiDelegationAction
...
+    deriving anyclass NFData
+
+instance ToJSON EncryptMetadataMethod where
+    toJSON AES256CBC = "basic"
+instance FromJSON EncryptMetadataMethod where
+    parseJSON = withText "base" $ \txt ->
+        if txt == "basic" then
+            pure AES256CBC
+        else
+            fail "'basic' is expected."
```
Comment (jonathanknowles) [inline lib/api/src/Cardano/Wallet/Api/Types.hs]: Suggestion: derive this instance automatically.

See: https://github.com/cardano-foundation/cardano-wallet/pull/4588

File: lib/api/src/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 4]
```diff
@@ -3121,6 +3153,96 @@ constructTransaction api argGenChange knownPools poolStatus apiWalletId body = d
...
+    (secretKey, iv) = PBKDF2.generateKey metadataPBKDF2Config pwd saltM
+    getMsgValue (Cardano.TxMetaText metaField, metaValue) =
+        if metaField == "msg" then
+            Just metaValue
+        else Nothing
+    getMsgValue _ = Nothing
+    merge Nothing (Just val) = Just val
+    merge (Just val) Nothing = Just val
+    merge Nothing Nothing = Nothing
+    merge (Just _) (Just _) = error "only one 'msg' field expected"
```
Comment (jonathanknowles) [inline lib/api/src/Cardano/Wallet/Api/Http/Shelley/Server.hs]: @paweljakubas is it possible to trigger this run-time error from the API?

I've created an experimental test to test this hypothesis here:
https://github.com/cardano-foundation/cardano-wallet/pull/4601

File: specifications/api/swagger.yaml [hunk 5]
```diff
@@ -1788,17 +1788,29 @@ x-transactionMetadata: &transactionMetadata
...
+    If used then metadata in a transaction is going to be encrypted by
+    AES 256 using CBC mode which is a default method (called base).
+    PBKDF2 password stretching is used to get a 32-byte secret key and a 16-byte
+    initialization vector required in the cipher.
+    PBKDF2 encryption algorithm using HMAC with the SHA256 hash algorithm is employed,
+    and 10000 iterations to get key and iv pair are used.
+    Cipher algorithm uses 8-byte salt, PKCS#7 padding as specified in
+    https://datatracker.ietf.org/doc/html/rfc5652#section-6.3 is applied.
+    Only metadata value under `msg` field is encrypted. If `msg` field is missing error
+    will be emitted.
```
Comment (jonathanknowles) [inline specifications/api/swagger.yaml:1807]: As a user who is new to this feature, I'm finding this `description` section a little hard to follow!

For example, it refers to a field called `msg`. However, this object only has two fields: `enc` and `passphrase`.

From reading through the tests (`Api.TypesSpec`), I can see that `msg` is supposed to be the key in a `TxMetaMap`.

Is there any way that we can give a more complete description of how to use this feature? (Ideally, so that the user doesn't have to go hunting through CIP 83.)

File: specifications/api/swagger.yaml [hunk 6]
```diff
@@ -4732,6 +4744,17 @@ x-errWithdrawalNotPossibleWithoutVote: &errWithdrawalNotPossibleWithoutVote
       type: string
       enum: ['withdrawal_not_possible_without_vote']
 
+x-errInvalidMetadataEncryption: &errInvalidMetadataEncryption
+  <<: *responsesErr
+  title: invalid_metadata_encryption
+  properties:
+    message:
+      type: string
+      description: The metadata has either wrong structure or cannot be encrypted.
```
Comment (jonathanknowles) [inline specifications/api/swagger.yaml]:
```suggestion
      description: |
          The supplied metadata object is not compatible with the format
          specified by CIP-83 (https://cips.cardano.org/cip/CIP-83).
```

**Review by jonathanknowles — APPROVED** [review]
Hi @paweljakubas 

Many thanks for creating this PR and for your patience with waiting for reviews.

Given this PR adds just the **_encryption_** functionality, with **_decryption_** planned for a later PR, I think this PR is okay to merge for now.

However, I do have some proposals that I think should be addressed in subsequent PRs:

1. Currently, the "business logic" for metadata encryption is located within the `Cardano.Wallet.Api.Http.Shelley.Server` module. I would propose to extract this logic out into a pure function that (a). doesn't rely on HTTP API types and (b). lives in a more appropriate location, preferably outside of the `Server` module, so that the `Server` module only needs to handle concerns that directly relate to the HTTP API. Perhaps this new location could be something like `Cardano.Write.Tx.Metadata.Encryption`?

2. I would propose to replace all instances of the magic value `674` with references to the `cip20MetadataKey` constant (already used by `toMetadataEncrypted`), and similarly, to re-locate this constant to the proposed `Cardano.Write.Tx.Metadata.Encryption` module.

3. When we have a pure function that performs **_decryption_**, I would propose to:
    1. add a new test module `Cardano.Write.Tx.Metadata.EncryptionSpec`.
    1. add a QuickCheck property test for the encryption-decryption round-trip property: `decrypt . encrypt == id`.
    1. add a selection of golden example-based unit tests (i.e., tests that are very quick to run that don't require us to fire up the integration test suite).

4. I still find the description of `x-encryptMetadata` in the swagger file a little hard to follow. Perhaps it would be possible to find someone who is completely new to this feature, and work with them to revise this description in a way that is easier for new users to understand? (See earlier review comment: https://github.com/cardano-foundation/cardano-wallet/pull/4555/files#r1599427620)

Again, many thanks again for creating this PR!

Jonathan

File: specifications/api/swagger.yaml (hunk 6 above)
Comment (paweljakubas) [reply specifications/api/swagger.yaml]: done

File: lib/api/src/Cardano/Wallet/Api/Http/Shelley/Server.hs (hunk 4 above)
Comment (jonathanknowles) [reply lib/api/src/Cardano/Wallet/Api/Http/Shelley/Server.hs]: Conclusion: I think it's impossible to trigger this call to `error` from the API.

This is because the JSON serialisation for `TxMetaMap` ensures that there can never be more than one map entry with any given key.

(It's unfortunate that this restriction is not reflected in the choice of type for `TxMetaMap`, which is just a list of pairs.)
